db/range_del_aggregator.cc

```diff
@@ -238,41 +268,47 @@ class CollapsedRangeDelMap : public RangeDelMap {
     abort();
   }
 
-  void AddTombstone(RangeTombstone t) override {
+  void AddTombstone(RangeTombstone t, bool is_truncated) override {
     if (ucmp_->Compare(t.start_key_, t.end_key_) >= 0 || t.seq_ == 0) {
       // The tombstone covers no keys. Nothing to do.
       return;
     }
 
     auto it = rep_.upper_bound(t.start_key_);
-    auto prev_seq = [&]() {
-      return it == rep_.begin() ? 0 : std::prev(it)->second;
+    auto prev_point_seq = [&]() {
+      return it == rep_.begin() ? 0 : std::prev(it)->second.point_seqnum;
+    };
+    auto prev_range_seq = [&]() {
+      return it == rep_.begin() ? 0 : std::prev(it)->second.range_seqnum;
     };
 
-    // end_seq stores the seqno of the last transition that the new tombstone
-    // covered. This is the seqno that we'll install if we need to insert a
-    // transition for the new tombstone's end key.
+    // end_seq stores the range seqno of the last transition that the new
+    // tombstone covered. This is the range seqno that we'll install if we need
+    // to insert a transition for the new tombstone's end key.
     SequenceNumber end_seq = 0;
 
-    // In the diagrams below, the new tombstone is always [c, k) @ 2. The
-    // existing tombstones are varied to depict different scenarios. Uppercase
-    // letters are used to indicate points that exist in the map, while
-    // lowercase letters are used to indicate points that do not exist in the
-    // map. The location of the iterator is marked with a caret; it may point
-    // off the end of the diagram to indicate that it is positioned at a
-    // entry with a larger key whose specific key is irrelevant.
-
-    if (t.seq_ > prev_seq()) {
+    // In the diagrams below, the new tombstone is [c, k) @ 2, or [c, k] @ 2 if
+    // specified. The existing tombstones are varied to depict different
+    // scenarios. Uppercase letters are used to indicate points that exist in
+    // the map, while lowercase letters are used to indicate points that do not
+    // exist in the map. The location of the iterator is marked with a caret;
+    // it may point off the end of the diagram to indicate that it is
+    // positioned at a entry with a larger key whose specific key is
+    // irrelevant. Truncated end keys not pointed at by the caret are marked by a *;
+    // those pointed at by the caret are marked by a !.
```

Comment:
Ignore this for now - I was in the process of commenting before the rebase so the comment improvements are in a weird half-finished state.